pkg/reconciler/common/image_override.go

```diff
@@ -0,0 +1,71 @@
+package common
+
+import (
+	"os"
+	"strings"
+
+	"knative.dev/operator/pkg/apis/operator/v1alpha1"
+)
+
+const imagePrefix = "IMAGE_"
+
+// imagesFromEnviron overrides registry images
+func imagesFromEnviron(obj v1alpha1.KComponent) {
+	reg := obj.GetSpec().GetRegistry()
+
+	reg.Override = imageMapFromEnvironment()
+
+	if defaultVal, ok := reg.Override["default"]; ok {
+		reg.Default = defaultVal
+	}
+
+	if ks, ok := obj.(*v1alpha1.KnativeServing); ok {
+		if qpVal, ok := reg.Override["queue-proxy"]; ok {
+			configure(ks, "deployment", "queueSidecarImage", qpVal)
+		}
+	}
+}
+
+func imageMapFromEnvironment() map[string]string {
+	overrideMap := map[string]string{}
+
+	for _, e := range os.Environ() {
+		pair := strings.SplitN(e, "=", 2)
+		if strings.HasPrefix(pair[0], imagePrefix) {
+			// convert
+			// "IMAGE_container=docker.io/foo"
+			// "IMAGE_deployment__container=docker.io/foo2"
+			// "IMAGE_env_var=docker.io/foo3"
+			// "IMAGE_deployment__env_var=docker.io/foo4"
+			// to
+			// container: docker.io/foo
+			// deployment/container: docker.io/foo2
+			// env_var: docker.io/foo3
+			// deployment/env_var: docker.io/foo4
+			name := strings.TrimPrefix(pair[0], imagePrefix)
+			name = strings.Replace(name, "__", "/", 1)
+			if pair[1] != "" {
```

Comment:
nit: you could do the check before extracting the name to save some unnecessary effort and immediately continue the loop . But a bit less nitty: Would it make sense to throw an error here as you _expect_ to have a value part for the environment ?

Reply:
In general it might make sense, yes. Practically though, there is no mechanism to surface this error to the reconcile process currently, which means that the patch area of this would be vastly increased and causing us a lot more trouble.

Seeing as we have E2E checks in place that make sure that no `gcr.io` based image makes it into our deployment, I think we at least have a fairly good idea of whether or not we butchered the list so I'm leaning towards leaving the failure mode as-is.

Reply:
ok, sounds reasonable.